Read script dependencies from generated files

sirreal opened the pull request:
Fixes #14837 

- Remove manually maintained static list of script dependencies.
- Read script dependencies from dependency files generated by build tools.
- Replace translation override usage of static list of scripts.
- Enable `injectPolyfill` option for the webpack plugin. Consumers of `wp-scripts` default webpack build will also receive this behavior.
- Remove addition of `wp-polyfill` dependency on script registration. It's handled by the webpack plugin.
- Include `build/[package]/*.deps.json` dependency files in plugin zip
- Handle undetected dependencies: `editor`, `media-models`, `media-views`, `postbox`.

## Description
Use generated script dependencies

## How has this been tested?

- Prepare local testing environment and `npm run build`. Does the block editor work as expected?
- Try building a plugin zip (`./bin/build-plugin-zip.sh`) then installing and activating it on another site. Does it work as expected?
  Sample from build plugin zip:
  ```
  adding: build/a11y/index.deps.json (stored 0%)
  adding: build/a11y/index.js (deflated 54%)
  adding: build/annotations/index.deps.json (deflated 20%)
  adding: build/annotations/index.js (deflated 67%)
  adding: build/api-fetch/index.deps.json (deflated 15%)
  adding: build/api-fetch/index.js (deflated 62%)
  adding: build/autop/index.deps.json (stored 0%)
  adding: build/autop/index.js (deflated 64%)
  ```

## Types of changes

Improve build tooling.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

sirreal committed:
Rework script registration to use generated deps

sirreal committed:
Deprecate static list of dependencies

gziolo commented:
We will need to expose `*.deps.json` files in the zip file for the plugin's package:
https://github.com/WordPress/gutenberg/blob/master/bin/build-plugin-zip.sh#L110

sirreal committed:
Rework translation without packages-dependencies.php

sirreal committed:
Include build/*/*.deps.json in plugin zip

sirreal committed:
Add script dependencies undetectable by build tools

Some scripts have dependencies that are undetectable by the webpack
plugin used to generate the dependency files. Add these dependencies
to the generated dependencies.

sirreal committed:
Add `wp-polyfill` dependency via webpack plugin

Enable `injectPolyfill` option in the webpack plugin instead of manually
injecting the dependency via PHP on script registration.

This will enable the polyfill for consumers of wp-scripts default
webpack config.

sirreal committed:
Improve deprecated message

sirreal commented on lib/packages-dependencies.php:
```diff
@@ -1,265 +1,5 @@
 /**
- * Returns array with the list of WordPress packages and their dependencies.
- *
+ * @deprecated since 5.6. Script dependencies should be read from generated `build/[package]/index.deps.json`.
```
I've had experience with plugins that avoid removing PHP files because it's caused problems with some hosts with unusual setups. Is that a concern here? Should this file be removed completely?

gziolo replied:
We never did it before. I think it's fine to remove.

sirreal replied:
Removed, that fixes one of the lint errors.

sirreal commented on packages/scripts/config/webpack.config.js:
```diff
@@ -67,7 +67,7 @@ const config = {
 		// when running watch mode.
 		! isProduction && new LiveReloadPlugin( { port: process.env.WP_LIVE_RELOAD_PORT || 35729 } ),
-		new DependencyExtractionWebpackPlugin(),
+		new DependencyExtractionWebpackPlugin( { injectPolyfill: true } ),
```
Includes `wp-polyfill` as a script dependency for every entrypoint.

aduth replied:
> Includes `wp-polyfill` as a script dependency for every entrypoint.

Looking at the implementation, would `wp-polyfill` include `wp-polyfill` as a dependency of itself?

sirreal replied:
It would yes.

That said, I don't see polyfill being built as part of Gutenberg (no `build/polyfill/index.deps.json`). Is this a theoretical concern or is the polyfill built somewhere we'd need to be careful with this?

gziolo replied:
I don't think we will ever scan `wp-polyfill` but we can add guard clause to ba on the safe side.

sirreal commented on lib/client-assets.php:
```diff
@@ -114,9 +114,14 @@ function gutenberg_override_translation_file( $file, $handle ) {
+
+	// Ignore scripts that are not found in the expected `build/` location.
+	$script_path = gutenberg_dir_path() . 'build/' . substr( $handle, 3 ) . '/index.js';
+	if ( ! file_exists( $script_path ) ) {
```
I'm unsure how to test whether this is working well, I'd appreciate some help.

gziolo replied:
That's a great question. Travis seems to be happy about changes applied 🤷‍♂️ 

How about we compare all script tags in the HTML output between `master` and this branch?

sirreal replied:
This is rather tricky to verify, but it _seems_ to be working. Here's what I did:

- Build the plugin zip from this PR.
- Get a new WordPress site.
- Install the plugin from the zip.
- Change the site language.
- Go to updates and install language updates. This will download the missing language files for the Gutenberg plugin we just uploaded.
- Open the editor.
- Check a known string: `wp.i18n.__( 'Use the classic WordPress editor.' )`
- Find the _plugin_ language file containing your string. In my case `/wp-content/languages/plugins/gutenberg-es_ES-d83116db7e3459d9c9a90cae3216278f.json`
- Edit the language file changing the _translation_ of your string (`Utiliza el editor clásico de WordPress.`) for something else (`👋 Hi from the plugin language file.`)
- Activate the plugin.
- Check the string from the editor again: `wp.i18n.__( 'Use the classic WordPress editor.' )`
- Should be your new string.
- Deactivate the plugin.
- Back to the original string.

sirreal commented:
Some tests are failing on CI, I'll have a look…

gziolo commented:
> Some tests are failing on CI, I'll have a look…

Skip them, those are tests which fail also for other PRs. I'm going to skip those tests later this week to avoid false negatives ...

gziolo commented on lib/client-assets.php:
```diff
@@ -177,17 +182,37 @@ function gutenberg_override_style( $handle, $src, $deps = array(), $ver = false,
+			: array();
+
+		// Add dependencies that cannot be detected and generated by build tools.
+		switch ( $handle ) {
```
Works for me 🤷‍♂️ 

😅

sirreal committed:
Remove obsolete lib/packages-dependencies.php

gziolo commented:
@omarreiss @atimmer - this is something we can add to the workflow in WordPress core as well. We probably will have to look at extending metadata included in this JSON file with the hash generated based on the content of build file to ensure it gets cached in the browser for as long as it doesn't change.

gziolo reviewed (approved):
It looks like this PR is ready to go. I tested it locally and it works as expected. Great work @sirreal!!!

This work to automate dependencies management will greatly improve the experience moving forward 💯

aduth commented on lib/client-assets.php:
```diff
@@ -177,17 +182,37 @@ function gutenberg_override_style( $handle, $src, $deps = array(), $ver = false,
  */
 function gutenberg_register_packages_scripts() {
-	$packages_dependencies = include dirname( __FILE__ ) . '/packages-dependencies.php';
+	foreach ( glob( gutenberg_dir_path() . 'build/*/index.js' ) as $path ) {
```
Depending if and how similar code would make its way to core, I'm a bit sensitive to the performance implications of a `glob`.

- https://gschoppe.com/wordpress/benchmarking-gutenberg/#comment-3154
- https://www.phparch.com/2010/04/putting-glob-to-the-test/

Seems like there could be optimizations:

- Avoiding the folder scan and pregenerating the scripts at build time (similar to [`generate-gutenberg-php.php`](https://github.com/WordPress/gutenberg/blob/master/bin/generate-gutenberg-php.php))
- Substituting `glob` with something faster (`opendir`)

sirreal replied:
👍A nice enhancement for a future version or simply another part of the build would be to produce what is needed in PHP without needing to hit the filesystem or parse JSON.

This came up during the core-js slack chat.

CC: @gziolo

gziolo replied:
Related conversation on Slack:

https://wordpress.slack.com/archives/C5UNMSU4R/p1556633084220800
